src/libraries/Common/src/System/NativeMemoryHelper.cs

```diff
@@ -0,0 +1,113 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.InteropServices;
+using System.Text;
+
+namespace System
+{
+    internal static unsafe class NativeMemoryHelper
```

Comment:
This should be called `NativeMemory` and only included for pre-.NET 6 targets. We do not want to have the wrapper on .NET 6.

Reply:
That can work for `Alloc` and `Free`, but it also has some replacements for `Marshal.StringToHGlobalUni` that would have to stay on a type like this. Is that okay?

Reply:
I would avoid touching StringToHGlobalUni in this PR. If it is useful to have NativeMemory variant of `StringToHGlobalUni` and similar APIs, we should introduce a proper public API for it first. Duplicating NativeMemory variant of `StringToHGlobalUni` via private helper is not an improvement over using StringToHGlobalUni directly.

Reply:
This might be worth profiling more and its probably worth opening a proposal.

On my box, an API which does simply the following, over a `260` character string:
```csharp
var ptr = Marshal.StringToHGlobalUni(s);
Marshal.FreeHGlobal(ptr);
```

Compared to an equivalent API using `NativeMarshal.Alloc/Free` under the covers:
```
BenchmarkDotNet=v0.13.0, OS=Windows 10.0.19043.1052 (21H1/May2021Update)
AMD Ryzen 9 5950X, 1 CPU, 32 logical and 16 physical cores
.NET SDK=6.0.100-preview.5.21302.13
  [Host]     : .NET 6.0.0 (6.0.21.30105), X64 RyuJIT
  DefaultJob : .NET 6.0.0 (6.0.21.30105), X64 RyuJIT

|  Method |     Mean |    Error |   StdDev |
|-------- |---------:|---------:|---------:|
|   Alloc | 41.92 ns | 0.051 ns | 0.045 ns |
| HGlobal | 47.37 ns | 0.071 ns | 0.066 ns |
```

Of course, in comparison to the GC transition and any work the call itself does, this is probably not that meaningful but ~10% overhead can add up.

Reply:
(To clarify, I don't think anything needs to be done in this PR, just calling out there is a measurable difference here and so it might be worth opening a proposal because of that. In more extreme cases, many `Alloc` calls can be up to 20x faster than many `LocalAlloc` calls.